File: LeetCode/2. Add Two Numbers/Solution.swift
/*

2. Add Two Numbers

You are given two non-empty linked lists representing two non-negative integers. The digits are stored in reverse order and each of their nodes contain a single digit. Add the two numbers and return it as a linked list.

You may assume the two numbers do not contain any leading zero, except the number 0 itself.

Example:

Input: (2 -> 4 -> 3) + (5 -> 6 -> 4)
Output: 7 -> 0 -> 8
Explanation: 342 + 465 = 807.

*/

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.next = nil
 *     }
 * }
 */

class Solution {
    func addTwoNumbers(_ l1: ListNode?, _ l2: ListNode?) -> ListNode? {
        var p = l1
        var q = l2
        var carry = 0
        var node: ListNode?
        var head: ListNode?

        while p != nil || q != nil {
            let value = (p?.val ?? 0) + (q?.val ?? 0) + carry
            let digit = value % 10
            carry = value / 10

            p = p?.next
            q = q?.next

            let newNode = ListNode(digit)
            node?.next = newNode
            node = newNode

            if head == nil {
                head = newNode
            }
        }

        if carry > 0 {
            let newNode = ListNode(carry)
            node?.next = newNode
            node = newNode
        }

        return head
    }
}

File: LeetCode/2. Add Two Numbers/Solution.cpp
/*

2. Add Two Numbers

You are given two non-empty linked lists representing two non-negative integers. The digits are stored in reverse order and each of their nodes contain a single digit. Add the two numbers and return it as a linked list.

You may assume the two numbers do not contain any leading zero, except the number 0 itself.

Example:

Input: (2 -> 4 -> 3) + (5 -> 6 -> 4)
Output: 7 -> 0 -> 8
Explanation: 342 + 465 = 807.

*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* head = new ListNode(0);
        ListNode* res = head;
        ListNode* p = l1;
        ListNode* q = l2;

        int carry = 0;

        while (p || q) {
            int p_num = p != nullptr ? p->val : 0;
            int q_num = q != nullptr ? q->val : 0;
            int sum = carry + p_num + q_num;

            carry = sum / 10;

            res->next = new ListNode(sum % 10);
            res = res->next;

            if (p != nullptr) {
                p = p->next;
            }

            if (q != nullptr) {
                q = q->next;
            }
        }

        if (carry > 0) {
            res->next = new ListNode(carry);
        }

        return head->next;
    }
};